pkg/server/plugin/keymanager/gcpkms/gcpkms.go

```diff
@@ -0,0 +1,1040 @@
+}
+
+// getPluginData gets the pluginData structure maintained by the plugin.
+func (p *Plugin) getPluginData() (*pluginData, error) {
+	p.pdMtx.RLock()
+	defer p.pdMtx.RUnlock()
+
+	if p.pd == nil {
+		return nil, status.Error(codes.FailedPrecondition, "plugin data not yet initialized")
+	}
+	return p.pd, nil
+}
+
+// setIamPolicy sets the IAM policy specified in the KeyPolicyFile to the given
+// resource. If there is no KeyPolicyFile specified, a default policy is constructed
+// and set to the resource.
+func (p *Plugin) setIamPolicy(ctx context.Context, cryptoKeyName string) (err error) {
+	log := p.log.With(cryptoKeyNameTag, cryptoKeyName)
+
+	// Get the handle to be able to inspect and change the policy of the
+	// CryptoKey.
+	h := p.kmsClient.ResourceIAM(cryptoKeyName)
+	if h == nil {
+		return errors.New("could not get Cloud KMS Handle")
+	}
+
+	// We use V3 for policies.
+	h3 := h.V3()
+	if h3 == nil {
+		return errors.New("could not get Cloud KMS Handle3")
+	}
+
+	// Get the policy.
+	policy, err := h3.Policy(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to retrieve IAM policy: %w", err)
+	}
+
+	// We expect the policy to be empty.
+	if len(policy.Bindings) > 0 {
+		// The policy is not empty, log the situation and do not replace it.
+		log.Warn("The CryptoKey already has a policy. No policy will be set.")
+		return nil
+	}
+	pd, err := p.getPluginData()
+	if err != nil {
+		return err
+	}
+
+	if pd.customPolicy != nil {
+		// There is a custom policy defined.
+		if err := h3.SetPolicy(ctx, pd.customPolicy); err != nil {
+			return fmt.Errorf("failed to set custom IAM policy: %w", err)
+		}
+		log.Debug("IAM policy updated to use custom policy")
+		return nil
+	}
+
+	// No custom policy defined. Build the default policy.
+	serviceAccount, err := p.getAuthenticatedServiceAccount()
+	if err != nil {
+		return status.Errorf(codes.Internal, "failed to get current identity: %v", err)
+	}
+	policy.Bindings = []*iampb.Binding{
+		{
+			Role:    "roles/cloudkms.signerVerifier",
+			Members: []string{fmt.Sprintf("serviceAccount:%s", serviceAccount)},
+		},
+	}
+	if err := h3.SetPolicy(ctx, policy); err != nil {
+		return fmt.Errorf("failed to set default IAM policy: %w", err)
+	}
+	log.Debug("IAM policy updated to use default policy")
+	return nil
+}
+
+// setPluginData sets the pluginData structure maintained by the plugin.
+func (p *Plugin) setPluginData(pd *pluginData) {
+	p.pdMtx.Lock()
+	defer p.pdMtx.Unlock()
+
+	p.pd = pd
+}
+
+// keepActiveCryptoKeys keeps CryptoKeys managed by this plugin active updating
+// the spire-last-update label with the current Unix time.
+func (p *Plugin) keepActiveCryptoKeys(ctx context.Context) error {
+	p.log.Debug("Keeping CryptoKeys managed by this server active")
+
+	p.entriesMtx.RLock()
+	defer p.entriesMtx.RUnlock()
+	var errs []string
+	for _, entry := range p.entries {
+		entry.cryptoKey.Labels[labelNameLastUpdate] = fmt.Sprint(p.hooks.clk.Now().Unix())
```

Comment:
this function has a RLock, but it is setting things here

Reply:
Good catch!